pkg/build/prometheus/metrics_via_collectors.go

```diff
@@ -0,0 +1,159 @@
+	StageDurationBucketQuery = BuildSubsystem + separator + StageDuration + separator + "bucket"
+	StageDurationCountQuery  = BuildSubsystem + separator + StageDuration + separator + "count"
+	StageDurationSumQuery    = BuildSubsystem + separator + StageDuration + separator + "sum"
+	StepDuration             = "steps_duration_seconds"
+	StepDurationBucketQuery  = BuildSubsystem + separator + StepDuration + separator + "bucket"
+	StepDurationCountQuery   = BuildSubsystem + separator + StepDuration + separator + "count"
+	StepDurationSumQuery     = BuildSubsystem + separator + StepDuration + separator + "sum"
+)
+
+var (
+	// decided not to have a separate counter for failed builds, which have reasons,
+	// vs. the other "finished" builds phases, where the reason is not set
+	descTerminalBuildCount = prometheus.NewDesc(
+		BuildSubsystem+separator+TerminalBuildCount,
+		"Counts total teriminal builds by namespace, name, phase, and in case of failures, reason",
+		[]string{"phase", "reason", "namespace", "name"},
+		nil,
+	)
+	descActiveBuildCount = prometheus.NewDesc(
+		BuildSubsystem+separator+ActiveBuildCount,
+		"Show current counts of active builds by namespace, name, phase, and with start time noted",
+		[]string{"phase", "namespace", "name", "startime"},
+		nil,
+	)
+	descBuildDuration = prometheus.NewDesc(
+		BuildSubsystem+separator+BuildDuration,
+		"Histogram of durations in seconds of terminated builds, by termination type",
+		[]string{"phase", "namespace", "name"},
+		nil,
+	)
+	descStageDuration = prometheus.NewDesc(
+		BuildSubsystem+separator+StageDuration,
+		"Histogram of durations in seconds of stages that have finished, by stage name",
+		[]string{"stage", "namespace", "name"},
+		nil,
+	)
+	descStepDuration = prometheus.NewDesc(
+		BuildSubsystem+separator+StepDuration,
+		"Histogram of durations in seconds of steps that have finished, by stage name and step name",
+		[]string{"stage", "step", "namespace", "name"},
+		nil,
+	)
+	bc         = buildCollector{}
+	registered = false
+)
+
+type buildCollector struct {
+	lister internalversion.BuildLister
+}
+
+// kube-state-metrics leverages prometheus.Registerer but does not exist in the version of prometheus
+// vendored into origin
+
+func IntializeMetricsCollector(buildLister internalversion.BuildLister) {
+	bc.lister = buildLister
+	// unit tests unearthed duplicate registration with prometheus via multiple calls to new build controller
+	if !registered {
+		prometheus.MustRegister(&bc)
+		registered = true
+	}
+	glog.V(4).Info("build metrics registered with prometheus")
+}
+
+// Describe implements the prometheus.Collector interface.
+func (bc *buildCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- descTerminalBuildCount
+	ch <- descActiveBuildCount
+	ch <- descBuildDuration
+	ch <- descStageDuration
+	ch <- descStepDuration
+}
+
+// Collect implements the prometheus.Collector interface.
+func (bc *buildCollector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
For 2, someone can make an alert on the histogram/max value, and then call the API and return the highest build.  Metrics that support alerting but require some slight API action are better than metrics that recreate API data, especially if it reduces cardinality.